File: src/dev/lib/doctext.ts
import { exec } from './shell';
import { curryWrap, curryTop } from './curry';

const regex_footnotemark = /\[\^(\d+)\]/g;
const regex_footnote = /\[\^(\d+)\]:\s*(.*)/g;

const supportedExtensions = ['md', 'docx', 'txt', 'pdf'];

const convertToMD = ({ filename }: { filename: string }) => async () => {
  const fileExt = filename.split('.').pop();
  if (fileExt !== undefined && !supportedExtensions.includes(fileExt))
    return Promise.reject(`Unsupported file type: ${fileExt}`);

  await exec`rm -rf temp`.catch(
    e => (e.includes('No such file or directory'))
      ? Promise.reject(e) : Promise.resolve());

  console.log(`Converting ${filename} to markdown...`);

  const res = await exec`pandoc -t markdown --extract-media temp ${filename}`;
  return res;
}

const mdToFile = ({ filename }: { filename: string }) => async (md: string) => {
  const fileExt = filename.split('.').pop();
  const newFilename = `${filename.split('.').slice(0, -1)[0]}-translated.${fileExt}`

  const buf = new Response(md);
  console.log(`Converting back to ${newFilename} ...`);
  await exec`pandoc -o ${newFilename} -f markdown < ${buf}`;

  await exec`rm -rf temp`;
}

const fileToMDString = curryWrap(
  convertToMD,
  mdToFile
);

const stringToParagraphs = () => async (md: string) =>
  md
    .split('\n\n')
    .map(p => p
      .split('\n')
      .map(s => s.trim())
      .join(' ')
    );

const paragraphsToString = () => async (md: string[]) =>
  md.join('\n\n')

const splitByParagraphs = curryWrap(
  stringToParagraphs,
  paragraphsToString
);

const stringToSentences = () => async (md: string) => {
  const mod = md.replaceAll('。', '. ');
  const res = await exec`pandoc -t markdown < ${new Response(mod)}`;
  return res.split('\n')
    .map(p => p.trim())
    .map(p => (p === '') ? '\n' : p);
}

const sentencesToString = () => async (md: string[]) =>
  md.join('\n');

const splitBySentences = curryWrap(
  stringToSentences,
  sentencesToString
);

const removeFootnotes = curryTop(
  () => async (md: string) => {
    console.log('Removing footnotes...');
    return md
      .split('\n\n')
      .filter(p => !p.match(regex_footnote))
      .map(p => p.replace(regex_footnotemark, ''))
      .join('\n\n');
  }
);

const clumpParagraphs = (chars: number) => async (paragraphs: string[]) =>
  paragraphs
    .reduce((acc, p) => {
      if (acc.length === 0)
        return [p];
      if ((acc[acc.length - 1].length + p.length) < chars)
        return [...acc.slice(0, -1), acc[acc.length - 1] + '\n\n' + p];
      return [...acc, p];
    }, [] as string[])

const unclumpParagraphs = () => async (clumped: string[]) =>
  clumped
    .map(p => p.split('\n\n'))
    .flat();

const clumpText = curryWrap(
  clumpParagraphs,
  unclumpParagraphs
);

export {
  fileToMDString,
  splitByParagraphs,
  splitBySentences,
  removeFootnotes,
  clumpText
}

File: src/dev/lib/doctext.py
# import fitz
import re
import subprocess

import more_itertools
import pdf2docx

doc_dir = "./doc/"

regex_footnotemark = r"\[\^(\d+)\]"
regex_footnote = r"\[\^(\d+)\]:\s*(.*)"

supported_extensions = {"md": "markdown", "docx": "docx", "txt": "plain", "pdf": "pdf"}


def is_nothing(text):
    return text == "" or text == "\n" or text == "\n\n" or text == "."


class Document:
    def __init__(self, filename, file_content=None):
        self.filename = filename
        self.file_ext = filename.split(".")[-1]
        if self.file_ext not in supported_extensions.keys():
            raise ValueError(f"File extension not supported: {self.file_ext}")

        subprocess.run(["mkdir", "-p", doc_dir])
        subprocess.run(["rm", "-rf", doc_dir + "temp"])

        # if (file_content is None)
        #     subprocess.run(["cp", "-rf", doc_dir + "temp"])
        # else
        open(doc_dir + filename, "wb").write(file_content)

        print(f"Converting {filename} to markdown...")
        res = ""
        if self.file_ext == "pdf":
            doc = fitz.open(filename)
            for page in doc:
                res += page.get_text("text") + "\n\n"
        else:
            res = subprocess.run(
                [
                    "pandoc",
                    "-t",
                    "markdown",
                    "--wrap=none",
                    "--extract-media",
                    doc_dir + "temp",
                    doc_dir + filename,
                ],
                stdout=subprocess.PIPE,
            ).stdout.decode("utf-8")
        #     print('Conversion to PDF is partially supported, changing to docx...')
        #     self.filename = self.filename.replace('.pdf', '.docx')
        #     pdf2docx.parse(filename, self.filename)
        #     self.file_ext = 'docx'
        # res = subprocess.run(['pandoc', '-t', 'markdown', '--wrap=none',
        #                       '--extract-media', 'temp', self.filename],
        #                      stdout=subprocess.PIPE).stdout.decode('utf-8')
        self.md = res
        self.split_options = None

    def __str__(self):
        return self.md

    def split(self, split_options="sentences"):
        self.split_options = split_options
        if self.split_options == "sentences":
            return string_to_sentences(self.md)
        elif self.split_options == "paragraphs":
            return string_to_paragraphs(self.md)
        else:
            raise ValueError(f"Invalid split option: {self.split_options}")

    def export(self, jobs, file_ext=None):
        if self.split_options == "sentences":
            self.md = sentences_to_string(jobs)
        elif self.split_options == "paragraphs":
            self.md = paragraphs_to_string(jobs)
        else:
            raise ValueError(f"Invalid split option: {self.split_options}")

        if file_ext is None:
            file_ext = self.file_ext
        if file_ext not in supported_extensions.keys():
            raise ValueError(f"File extension not supported: {file_ext}")
        if file_ext == "pdf":
            print("Conversion to PDF is partially supported, changing to docx...")
            file_ext = "docx"

        new_filename = self.filename.replace(
            f".{self.file_ext}", f"-translated.{file_ext}"
        )

        print(f"Converting back to {new_filename}...")
        subprocess.run(
            [
                "pandoc",
                "-o",
                doc_dir + new_filename,
                "-t",
                supported_extensions[file_ext],
                "-f",
                "markdown",
            ],
            input=self.md.encode("utf-8"),
        )

        subprocess.run(["rm", "-rf", doc_dir + "temp"])
        return (doc_dir, new_filename)


def string_to_paragraphs(md):
    paragraphs = md.split("\n\n")
    paragraphs_clean = list(map(lambda p: " ".join(p.split("\n")), paragraphs))
    return paragraphs_clean


def paragraphs_to_string(paragraphs):
    return "\n\n".join(paragraphs)


def string_to_sentences(md):
    fix_dots = md.replace("。", ". ")
    paragraphs = fix_dots.split("\n\n")
    sentences = list(map(lambda p: p.split(". "), paragraphs))
    sentences_clean = list(
        map(lambda p: list(filter(lambda s: not is_nothing(s), p)), sentences)
    )
    sentences_dot = list(
        map(lambda p: list(map(lambda s: s + ".", p)) + [""], sentences_clean)
    )
    sentences_str = list(more_itertools.collapse(sentences_dot))
    return sentences_str


def sentences_to_string(sentences):
    return "\n".join(sentences)


def remove_footnotes_func(_):
    def f(md):
        paragraphs = md.split("\n\n")
        paragraphs_no_footnotes = filter(
            lambda p: not re.compile(regex_footnote).match(p), paragraphs
        )
        paragraphs_no_footnotes = list(
            map(lambda p: re.sub(regex_footnotemark, "", p), paragraphs_no_footnotes)
        )
        str_no_footnotes = "\n\n".join(paragraphs_no_footnotes)
        return str_no_footnotes

    return f
